### PR TITLE
nfts by wallet

### PR DESCRIPTION
**Pull request by imabdulbasit**
Query NFTs by wallet address

### CODE REVIEW

File: crates/graphql/src/schema.rs [hunk 1]
```diff
@@ -961,6 +961,31 @@ impl QueryRoot {
 
         rows.pop().map(Into::into)
     }
+
+    fn nfts_by_wallet(&self, context: &AppContext, owner_address: String) -> Vec<Nft> {
```
Comment (kespinola) [inline crates/graphql/src/schema.rs:965]: Ideally I'd like this to be a filter in the existing nft query. That maybe be challenging but give it a look and let me know. nfts(owners: []) should be additive to the existing creators filter

**Review by kespinola — CHANGES_REQUESTED** [review]

File: crates/graphql/src/schema.rs (hunk 1 above)
Comment (imabdulbasit) [reply crates/graphql/src/schema.rs]: I have made creators and owners as optional parameters

Comment (imabdulbasit) [reply crates/graphql/src/schema.rs]: But the nfts query returns all the nfts when no parameter is passed , is that okay or do you want to return empty results when no parameter is passed

File: crates/graphql/src/schema.rs [hunk 2]
```diff
@@ -897,9 +897,13 @@ impl QueryRoot {
     fn nfts(
         &self,
         context: &AppContext,
-        #[graphql(description = "Filter on creator address")] creators: Vec<String>,
+        #[graphql(description = "Filter on owner address")] owners: Option<Vec<String>>,
+        #[graphql(description = "Filter on creator address")] creators: Option<Vec<String>>,
         #[graphql(description = "Filter on attributes")] attributes: Option<Vec<AttributeFilter>>,
     ) -> Vec<Nft> {
+        if owners.is_none() && creators.is_none() && attributes.is_none() {
+            return Vec::new();
```
Comment (kespinola) [inline crates/graphql/src/schema.rs:905]: Lets return a graphql api error here instead. I haven't looked that up yet.

**Review by kespinola — CHANGES_REQUESTED** [review]
Looks great. Just the change to a graphql api error. something like one of the following filters is required...

**Comment by kespinola** [comment]
I'll get the api error change in now.

File: crates/graphql/src/schema.rs (hunk 2 above)
Comment (imabdulbasit) [reply crates/graphql/src/schema.rs]: added

Comment (imabdulbasit) [reply crates/graphql/src/schema.rs]: image

File: crates/graphql/src/schema.rs [hunk 3]
```diff
@@ -919,12 +926,20 @@ impl QueryRoot {
...
+                    .eq(any(owners.unwrap_or_else(Vec::new)))
+                    .and(token_accounts::amount.eq(1)),
+            );
+        let metadata_addresses_subquery = metadata_creators::table
+            .select(metadata_creators::metadata_address)
+            .filter(metadata_creators::creator_address.eq(any(creators.unwrap_or_else(Vec::new))));
+
+        let rows: Vec<models::Nft> = query
+            .filter(metadatas::address.eq(any(metadata_addresses_subquery)))
+            .or_filter(metadatas::mint_address.eq(any(mint_addresses_subquery)))
```
Comment (kespinola) [inline crates/graphql/src/schema.rs:942]: Can some folks audit me on this? I switched 2 subqueries 1 for finding metadata addresses associated to the creators passed in and another for mint_address for owners passed in.

Then use a WHERE address IN (creators_metadata_addressses) OR mint_address IN (mint_addresses)

This query should return metadata that match the attribute filters and either the NFT is created by one of the creators or owned by one of the owners.

**Review by kespinola — APPROVED** [review]

File: crates/graphql/src/schema.rs (hunk 3 above)
Comment (imabdulbasit) [reply crates/graphql/src/schema.rs:942]: should work

File: crates/graphql/src/schema.rs [hunk 4]
```diff
@@ -919,11 +926,22 @@ impl QueryRoot {
             },
         );
 
+        let mint_addresses_subquery = token_accounts::table
+            .select(token_accounts::mint_address)
+            .filter(
+                token_accounts::owner_address
+                    .eq(any(owners.unwrap_or_else(Vec::new)))
+                    .and(token_accounts::amount.eq(1)),
```
Comment (kespinola) [inline crates/graphql/src/schema.rs:934]: A great index candidate [mint_address, amount === 1]

File: crates/graphql/src/schema.rs [hunk 5]
```diff
@@ -919,11 +926,22 @@ impl QueryRoot {
...
+            .filter(metadata_creators::creator_address.eq(any(creators.unwrap_or_else(Vec::new))));
+
         let rows: Vec<models::Nft> = query
             .filter(
-                metadatas::address.eq(any(metadata_creators::table
-                    .select(metadata_creators::metadata_address)
-                    .filter(metadata_creators::creator_address.eq(any(creators))))),
+                metadatas::address
+                    .eq(any(metadata_addresses_subquery))
+                    .or(metadatas::mint_address.eq(any(mint_addresses_subquery))),
```
Comment (kespinola) [inline crates/graphql/src/schema.rs:944]: Another index for mint_address on metadatas table.

File: crates/graphql/src/schema.rs [hunk 6]
```diff
@@ -919,11 +926,22 @@ impl QueryRoot {
...
+        let mint_addresses_subquery = token_accounts::table
+            .select(token_accounts::mint_address)
+            .filter(
+                token_accounts::owner_address
+                    .eq(any(owners.unwrap_or_else(Vec::new)))
+                    .and(token_accounts::amount.eq(1)),
+            );
+        let metadata_addresses_subquery = metadata_creators::table
+            .select(metadata_creators::metadata_address)
+            .filter(metadata_creators::creator_address.eq(any(creators.unwrap_or_else(Vec::new))));
```
Comment (kespinola) [inline crates/graphql/src/schema.rs:938]: Index on creator for metadata_creators.

**Review by kespinola — COMMENTED** [review]
notes on indexes to add.

File: crates/graphql/src/schema.rs [hunk 7]
```diff
@@ -897,13 +897,20 @@ impl QueryRoot {
...
         context: &AppContext,
-        #[graphql(description = "Filter on creator address")] creators: Vec<String>,
+        #[graphql(description = "Filter on owner address")] owners: Option<Vec<String>>,
+        #[graphql(description = "Filter on creator address")] creators: Option<Vec<String>>,
         #[graphql(description = "Filter on attributes")] attributes: Option<Vec<AttributeFilter>>,
-    ) -> Vec<Nft> {
+    ) -> FieldResult<Vec<Nft>> {
+        if owners.is_none() && creators.is_none() {
+            return Err(FieldError::new(
+                "No filter provided! Please provide atleast one of the filter",
```
Comment (ray-kast) [inline crates/graphql/src/schema.rs:906]:
```suggestion
                "No filter provided! Please provide at least one of the filters",
```

**Review by ray-kast — CHANGES_REQUESTED** [review]
Looks good

File: crates/graphql/src/schema.rs [hunk 8]
```diff
@@ -941,7 +959,7 @@ impl QueryRoot {
             .load(&conn)
             .unwrap();
 
-        rows.into_iter().map(Into::into).collect()
+        Ok(rows.into_iter().map(Into::into).collect())
```
Comment (ray-kast) [inline crates/graphql/src/schema.rs:962]: This function calls `.unwrap()` in several spots, so if the return type is to be changed to Result the panics should also be removed.

**Review by ray-kast — CHANGES_REQUESTED** [review]

File: crates/graphql/src/schema.rs (hunk 7 above)
Comment (ray-kast) [reply crates/graphql/src/schema.rs]: @imabdulbasit i see you resolved this, did the fix get committed?

File: crates/graphql/src/schema.rs (hunk 5 above)
Comment (imabdulbasit) [reply crates/graphql/src/schema.rs:944]: this index is already added added

File: crates/graphql/src/schema.rs (hunk 7 above)
Comment (imabdulbasit) [reply crates/graphql/src/schema.rs]: I pushed a commit

File: crates/graphql/src/schema.rs (hunk 8 above)
Comment (imabdulbasit) [reply crates/graphql/src/schema.rs:962]: yeah this needs to be changed in whole schema.rs

Comment (ray-kast) [reply crates/graphql/src/schema.rs:962]: Right.  For now I'd suggest just removing the FieldResult if possible and unwrapping.

File: crates/graphql/src/schema.rs (hunk 7 above)
Comment (ray-kast) [reply crates/graphql/src/schema.rs]: Changeset doesn't show it was fixed, are you sure?

Comment (imabdulbasit) [reply crates/graphql/src/schema.rs]: https://github.com/holaplex/solana-indexer/pull/128/commits/be6664be3d22cbf6f4b61731ef4b7d3780118467

Comment (ray-kast) [reply crates/graphql/src/schema.rs]: :+1: aight idk what happened on github's end lol

File: crates/graphql/src/schema.rs (hunk 8 above)
Comment (imabdulbasit) [reply crates/graphql/src/schema.rs:963]: I have changed FieldResult to Result<Vec<Nft>,FieldError>

Comment (ray-kast) [reply crates/graphql/src/schema.rs:963]: I think having FieldResult was fine, as long as all unwraps are changed to the `?` operator.

**Comment by ray-kast** [comment]
@imabdulbasit you'll need to rebase onto dev

**Review by ray-kast — APPROVED** [review]

**Pull request merged**